lib/internal/test_runner/mock/mock_timers.js

```diff
@@ -0,0 +1,183 @@
+'use strict';
+
+const {
+  emitExperimentalWarning,
+} = require('internal/util');
+
+const {
+  DateNow,
+  FunctionPrototypeBind,
+  globalThis,
+  Promise,
+} = primordials;
+
+const PriorityQueue = require('internal/priority_queue');
+const nodeTimers = require('timers');
+const nodeTimersPromises = require('timers/promises');
+const EventEmitter = require('events');
+
+function compareTimersLists(a, b) {
+  return (a.runAt - b.runAt) || (a.id - b.id);
+}
+
+function setPosition(node, pos) {
+  node.priorityQueuePosition = pos;
+}
+
+
+class MockTimers {
+  #realSetTimeout;
+  #realClearTimeout;
+  #realSetInterval;
+  #realClearInterval;
+
+  #realPromisifiedSetTimeout;
+  #realPromisifiedSetInterval;
+
+  #realTimersSetTimeout;
+  #realTimersClearTimeout;
+  #realTimersSetInterval;
+  #realTimersClearInterval;
+
+  #isEnabled = false;
+  #currentTimer = 1;
+  #now = DateNow();
+
+  #timers = new PriorityQueue(compareTimersLists, setPosition);
+
+  #setTimeout = FunctionPrototypeBind(this.#createTimer, this, false);
+  #clearTimeout = FunctionPrototypeBind(this.#clearTimer, this);
+  #setInterval = FunctionPrototypeBind(this.#createTimer, this, true);
+  #clearInterval = FunctionPrototypeBind(this.#clearTimer, this);
+
+  constructor() {
+    emitExperimentalWarning('The MockTimers API');
+  }
+
+  #createTimer(isInterval, callback, delay, ...args) {
+    const timerId = this.#currentTimer++;
+    this.#timers.insert({
+      __proto__: null,
+      id: timerId,
+      callback,
+      runAt: DateNow() + delay,
+      interval: isInterval,
+      args,
+    });
+
+    return timerId;
+  }
+
+  #clearTimer(position) {
+    this.#timers.removeAt(position);
+  }
+
+  async * #setIntervalPromisified(interval, startTime, ...args) {
+    const intervalEmitter = new EventEmitter();
+    const callback = () => intervalEmitter.emit('data', startTime);
+    this.#createTimer(true, callback, interval, ...args);
+
+    for await (const events of EventEmitter.on(intervalEmitter, 'data')) {
+      yield events.at(0);
+    }
+  }
+
+  tick(time = 1) {
+    // if (!this.isEnabled) {
+    //   throw new Error('you should enable fakeTimers first by calling the .enable function');
+    // }
+
+    // Increase the current time by the specified time.
+    this.#now += time;
+
+    // Execute all timers whose runAt time is less than or equal to the current time.
+    let timer = this.#timers.peek();
+    while (timer && timer.runAt <= this.#now) {
+      // Execute the timer's callback function with the specified arguments.
+      timer.callback(...timer.args);
```

Comment:
Can we avoid the unsafe array iteration please?

Reply:
what's unsafe on this iteration?